Add Version and Platform Checks section

Solumin opened the pull request:
Version check requirements drawn from PEP 484 and stub files in typeshed.

Solumin committed:
[WIP] Add Version and Platform Checks section

Version check requirements drawn from PEP 484 and stub files in typeshed.

Solumin commented:
Version checks are listed under "What constructs are typically supported by type checkers" on our wiki, but based on the PEP so far, the actual content goes in the Style Guide section. Let me know if it should be moved.

JelleZijlstra commented on pep-9999.rst:
```diff
@@ -259,6 +259,73 @@ No::
+    else:
+      # Python 2-specific type hints.
+
+    if sys.version_info.major >= 3:
```
Does this currently work? I think I prefer that we allow *only* tuple-based comparisons, so that type checkers don't have to support tuple and `.major`-based operations.

Solumin replied:
> Does this currently work?

Depends on the type checker, I suppose, but it's part of PEP 484: https://www.python.org/dev/peps/pep-0484/#version-and-platform-checking

>I think I prefer that we allow _only_ tuple-based comparisons, so that type checkers don't have to support tuple and `.major`-based operations.

I'm fine with this, personally. I think tuple-based comparisons are better overall, and I've opened #26 so we can discuss this.

Solumin replied:
Added recommendation for tuple-based comparisons.

JelleZijlstra commented on pep-9999.rst:
```diff
@@ -259,6 +259,73 @@ No::
+    if sys.version_info.major != 2:
+      # This is less clear than ">= 3".
+
+    if sys.version_info >= (3, 5, 1):
```
We should provide guidance on what  to do instead (always write stubs for the latest patch version).

Solumin replied:
I agree, but I'm not sure that belongs in this section.

srittau commented on pep-9999.rst:
```diff
@@ -259,6 +259,73 @@ No::
+Type stubs for libraries that support multiple Python versions can use version
+checks to supply version-specific type hints. Type stubs for different Python
+versions should still conform to the most recent Python version's syntax, as
+explain in the Syntax_ section above.
```
explained

Solumin replied:
Makes sense to me. Done.

srittau reviewed (dismissed):
Thank you for writing this. One general note: I have indented examples with four spaces (per PEP 8). We should stay consistent.

On a broader note: Maybe we should split this into two: Explain what kind of version checks are allowed in the "Supported constructs" section, but make a recommendation for using the tuple-based syntax here.

Solumin committed:
Split Version Check info into Supported Constructs and Style Guide.

Type stub files as per PEP 484 can have version checks that compare e.g. `sys.version_info.major >= 3`, but the tuple-based syntax is preferred.

Also fixes indentation in examples.

Solumin commented:
As per discussion here and on #26, I've updated the content of this PR to recommend tuple-based version checks. This resolves #26.

Solumin committed:
Remove extraneous line.

rchen152 reviewed (approved):
This looks reasonable to me. I don't like the fact that PEP 484 says that stubs should support things like `sys.version_info[0]`, but unless it's possible to change the PEP, I think the way this PR handles it is good.

Solumin committed:
Merge branch 'type-stub-pep' into version-checks-style-1

srittau commented on pep-9999.rst:
```diff
@@ -514,6 +555,61 @@ No::
+        # Not necessary since Python 3.3.
+
+    if sys.platform in ['linux', 'cygwin', 'darwin']:
+        # Only '==' should be used in platform checks.
```
I think "no" part can actually be moved to the first part. I don't think type checkers would support those constructs (i.e. everything more complex that `==` or `!=` and logical operators).

srittau reviewed (commented):
Looks good, just one comment below.

Solumin commented on pep-9999.rst:
```diff
@@ -149,6 +149,47 @@ Two kinds of structured comments are accepted:
+    else:
+        # Python 2-specific type hints.
+
+    if sys.version_info.major >= 3:
```
We don't want to encourage this syntax, so reword this.

rchen152 replied:
from the meeting, stubs should not support: version_info.major, .startswith and `in` checks

Solumin committed:
Move bad system check examples.

- moves the examples of discouraged system checks to the Supported Constructs section.
- reword the version check part of Support Constructs to actively discourage .major checks.

Solumin commented:
Alright, I've incorporated the changes from the planning meeting.

Solumin committed:
Fix typo.

rchen152 commented on pep-9999.rst:
```diff
@@ -149,6 +149,61 @@ Two kinds of structured comments are accepted:
+
+Type stubs for libraries that support multiple Python versions can use version
+checks to supply version-specific type hints. Type stubs for different Python
+versions should still conform to the most recent Python version's syntax, as
```
"should still conform to the most recent *supported* Python version's syntax", maybe? since we decided to change the syntax section

rchen152 commented on pep-9999.rst:
```diff
@@ -514,6 +569,53 @@ No::
+    if sys.version_info >= (3, 5, 1):
+        # Version checks should only check major and minor version numbers.
+
+    if sys.version_info[0] >= 3 and sys.version_info[1] >= 5:
```
These two examples may not be needed anymore, if we're not supporting the syntax.

Solumin replied:
The two sections serve slightly different purposes. Supported Constructs says type checkers don't need to support .major. The Style Guide says type stub authors should use tuple checks and to avoid per-element/.major checks. I think having both is important.

rchen152 replied:
That's fair.

Solumin committed:
Update w.r.t. supported python syntax
